app/models/metasploit/cache/module/ancestor.rb

```diff
@@ -244,26 +138,145 @@ class Metasploit::Cache::Module::Ancestor < ActiveRecord::Base
+    if block_given?
+      progress_bar.total = relative_paths.length
+
+      ActiveRecord::Base.connection_pool.with_connection do
+        updatable_module_ancestors = scope.where(relative_path: relative_paths)
+        new_relative_path_set = Set.new relative_paths
+
+        # use find_each since this is expected to exceed default batch size of 1000 records.
+        updatable_module_ancestors.find_each do |updatable_module_ancestor|
+          new_relative_path_set.delete(updatable_module_ancestor.relative_path)
+
+          changed = assume_changed
+
+          # real_path_modified_at and real_path_sha1_hex_digest should be updated even if assume_changed is true so
+          # that database stays in-sync with file system
+
+          updatable_module_ancestor.real_path_modified_at = updatable_module_ancestor.derived_real_path_modified_at
+
+          # only derive the SHA1 Hex Digest if modification time has changed to save time
+          if updatable_module_ancestor.real_path_modified_at_changed?
+            updatable_module_ancestor.real_path_sha1_hex_digest = updatable_module_ancestor.derived_real_path_sha1_hex_digest
+
+            changed ||= updatable_module_ancestor.real_path_sha1_hex_digest_changed?
+          end
+
+          if changed
+            yield updatable_module_ancestor
+            progress_bar.increment
+          else
+            # increment even when no yield so that increment occurs for each path and matches totally without jumps
+            progress_bar.increment
+          end
+        end
+
+        # after all pre-existing relative_paths are subtracted, new_relative_path_set contains only relative_paths not
+        # in the database
+        new_relative_path_set.each do |relative_path|
+          new_module_ancestor = scope.new(relative_path: relative_path)
+
+          yield new_module_ancestor
+          progress_bar.increment
+        end
+      end
+    else
+      enum_for(
+          __method__,
+          assume_changed: assume_changed,
+          progress_bar: progress_bar,
+          relative_paths: relative_paths,
+          scope: scope
+      )
+    end
+  end
+
+  # Ensure that only {#module_type} matching `MODULE_TYPE` is valid for `subclass`.
+  #
+  # @param subclass [Class<Metasploit::Cache::Module::Ancestor>] a subclass of {Metasploit::Cache::Module::Ancestor}.
+  # @return [void]
+  def self.restrict(subclass)
+    #
+    # Validations
+    #
 
-    handled = false
-    module_type = options[:module_type]
-    payload_type = options[:payload_type]
+    subclass.validate :module_type_matches
 
-    if module_type == 'payload' and HANDLED_TYPES.include? payload_type
-      handled = true
+    #
+    # Class Methods
+    #
+
+    subclass_relative_path_prefix = "#{subclass::MODULE_TYPE_DIRECTORY}".freeze
+
+    subclass.define_singleton_method(:relative_path_prefix) do
```

Comment:
I think this is a case of not going back and refactoring.  Originally, I was passing in the directory as an argument to restrict after reading an article about favoring explicit dsl arguments over magic includes, but later I added the constants for other reasons and ended up making the DSL magic again.  I'll change this to drop the DSL and favor a simple extend of a Module with these methods.